rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,75 @@
+
+Lack of determinism prevents companies from launching products using models developed in TF. For a subset of these industries having deterministic behavior is a regulatory requirement. 
+
+In addition, it increases model velocity development by reducing noise, while also simplifying the debugging workflow.
+
+## Design Proposal
+We will create a new flag with the default value of ‘False’ which enables determinism.  We will define 2 functions:
+
+* `tf.config.enable_deterministic_execution(enabled)`
+* `tf.config.deterministic_execution_enabled()`
+
+The first function takes in a boolean value, and allows the model developer to enable/disable determinism. The second function returns a bool indicating whether determinism is enabled.
+In some cases, we have deterministic and non-deterministic versions of the kernel. In such cases, we will use this flag to run the appropriate kernels.
+For ops which do not yet have a deterministic implementation, TensorFlow will raise an error `tf.errors.UnimplementedError` if the flag is enabled.
+
+Enabling deterministic execution does not automatically cause a user’s program to become deterministic. If users use non-deterministic constructs outside TensorFlow, such as threads/process, in ways that influence TensorFlow’s behavior, their program will not be deterministic. In order for a user to ensure their program is deterministic, users must both enable deterministic execution within TensorFlow and remove any sources of non-determinism outside TensorFlow.
+
+### Existing Flags
+Multiple environmental variables exist today that control determinism. As part of this change, we will deprecate then remove the following:
+
+* TF_DETERMINISTIC_OPS
+* TF_CUDNN_DETERMINISTIC
+
+tf.data also has flags for determinism. The system will throw an error message if flags are out of sync i.e. if deterministic_execution_enabled is enabled but if the tf.data option is set to ‘false’, we will throw an error. (`tf.data.Options.experimental_deterministic`). We’ll also add the necessary checks for Dataset.map and Dataset.interleave.
+
+### Grappler changes
+Grappler graph optimizations may add non-deterministic behavior. In particular some optimizations will time out if they take too long to run. When determinism is enabled, these time outs will be disabled.
+
+### Random ops
+Legacy random ops, such as `tf.random.normal`, are not deterministic if no seed is set, and so such ops will raise an error when determinism is enabled. To fix, the user should set a global seed with `tf.random.set_seed`. Since most models use legacy random ops, in practice users must call `tf.random.set_seed` when enabling deterministic behavior. Alternatively, users can pass a seed to every individual random operation, but doing so is more inconvenient.
```

Comment:
I have a note that `tf.image.sample_distorted_bounding_box` is an unusual op that has a seed parameter in that even if you call `tf.random.set_seed` it will still operate nondeterministically unless you set its seed parameter.